toolkit/tools/precacher/precacher.go

```diff
@@ -0,0 +1,323 @@
+
+	prof, err := profile.StartProfiling(profFlags)
+	if err != nil {
+		logger.Log.Warnf("Could not start profiling: %s", err)
+	}
+	defer prof.StopProfiler()
+
+	timestamp.BeginTiming("precacher", *timestampFile)
+	defer timestamp.CompleteTiming()
+
+	// Get the state of the current repo from a snapshot
+	rpmSnapshot, err := rpmSnapshotFromFile(*snapshot)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	// See what is available online from the upstream repos
+	availablePackages, err := getAllRepoData(*repoUrls)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	logger.Log.Infof("Found %d packages to use for pre-caching", len(availablePackages))
+	for _, pkg := range availablePackages {
+		logger.Log.Debugf("Found package: %s", pkg)
+	}
+
+	// For each package in the snapshot, check if it is available online and try to download it
+	downloadedPackages, err := downloadMissingPackages(rpmSnapshot, availablePackages, *outDir, *concurrentNetOps)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+
+	logger.Log.Infof("Pre-caching complete: Downloaded %d packages into the cache", len(downloadedPackages))
+	err = writeSummaryFile(*outputSummaryFile, downloadedPackages)
+	if err != nil {
+		logger.PanicOnError(err)
+	}
+}
+
+func rpmSnapshotFromFile(snapshotFile string) (rpmSnapshot *repocloner.RepoContents, err error) {
+	err = jsonutils.ReadJSONFile(snapshotFile, &rpmSnapshot)
+	return
+}
+
+func getAllRepoData(repoUrls []string) (namesToUrls map[string]string, err error) {
+	timestamp.StartEvent("pull available package data from repos", nil)
+	defer timestamp.StopEvent(nil)
+	// repoquery behaves differently on Mariner and Ubuntu
+	isMariner := shell.IsMarinerOs()
+
+	namesToUrls = make(map[string]string)
+	for _, repoUrl := range repoUrls {
+		packages, err := getRepoPackages(repoUrl, isMariner)
+		if err != nil {
+			return nil, err
+		}
+
+		// We will be searching by the name: "<name>-<version>.<distro>.<arch>", the results from the repoquery will be
+		// in the form of "(BASE_URL)/<name>-<version>.<distro>.<arch>.rpm"
+		for _, pkgUrl := range packages {
+			name := path.Base(pkgUrl)
+			name = strings.TrimSuffix(name, ".rpm")
+
+			// We need to prepend the repoUrl to the name to get the full url on Mariner
+			if isMariner {
+				pkgUrl = fmt.Sprintf("%s/%s", repoUrl, pkgUrl)
+			}
+			namesToUrls[name] = pkgUrl
+		}
+	}
+	return
+}
+
+// getRepoPackages returns a list of packages available in the given repoUrl by running repoquery
+func getRepoPackages(repoUrl string, isMariner bool) (packages []string, err error) {
+	const (
+		reqoqueryTool    = "repoquery"
+		randomNameLength = 10
+	)
+	var (
+		queryCommonArgList = []string{"-a", "--qf", "%{location}"}
+		marinerArgList     = []string{"-y", "-q", "--disablerepo=*"}
+		ubuntuArgList      = []string{"--show-duplicates", "--tempcache"}
+
+		finalArgList []string
+	)
+
+	logger.Log.Infof("Getting pre-caching package data from %s", repoUrl)
+	if isMariner {
```

Comment:
I've moved this to a chroot. I was hoping to avoid it since it takes ~5-10 seconds to build up the chroot and the way this tool runs it will always be invoked even if there is nothing to build, so it's a sunk cost for every build with PRECACHE=y set, but... I think it is the right approach. This will isolate us from any host specifics.

Reply:
Considering that this is a new feature, IMHO I don't think it would be too problematic to say it requires Ubuntu 22.04 to use.